### ISSUE TITLE
[Ch 17][Notebook][Minor] 삽질 코너 힌트 "shape + dtype 두 에러가 동시에 나 메시지가 길어진다" — 실제는 shape 한 줄, Ch 13 #108 과 같은 유형

### ISSUE THREAD

**Issue by jwrim9305** [yoon-gu/neuqes-101#276]
## 대상

- **챕터**: Ch 17 — 한국어 BERT Multi-label
- **산출물**: Notebook `17_ko_multilabel.ipynb` c42 (🚀 삽질 코너) · `pages/17-ko_multilabel-wrapup.md`
- **기준 커밋**: `5629f80` (master)
- **Colab T4 실제 실행 1회 수행** (transformers 5.16.1 · torch 2.11.0+cu128)

## 실제 결과

삽질 코너의 힌트는 이렇게 예고합니다.

> 힌트: `BCEWithLogitsLoss` 는 *logits 와 같은 shape 의 float 텐서* 를 라벨로 받는데, 위 코드는 *(B,) int* 를 넘깁니다. **shape mismatch + dtype mismatch 두 가지 에러가 동시에 날 수 있어 메시지가 길어집니다** (Ch 13 의 삽질과 같은 함정).

**실제로 돌려보면 단일 `ValueError` 한 줄입니다.**

```
예외 종류 : ValueError
메시지    : Target size (torch.Size([8])) must be the same as input size (torch.Size([8, 7]))
발생 지점 : torch/nn/functional.py:3638  binary_cross_entropy_with_logits
compute_loss 도달 : True
```

| 힌트의 예고 | 실제 |
|---|---|
| shape mismatch **+ dtype mismatch 두 가지** | **shape 하나만** |
| 에러 메시지가 **길어집니다** | **한 줄** (위 인용 전체) |

`dtype mismatch` 는 **나타나지 않습니다.** `F.binary_cross_entropy_with_logits` 가 shape 을 먼저 검사해 그 자리에서 멈추므로, dtype 검사까지 가지 않습니다. shape 을 맞춰 주면 그때 dtype 문제가 따로 드러나지만, 이 코드로는 도달하지 못합니다.

## 왜 이게 문제인가

삽질 코너는 **독자가 직접 돌려 보고 에러 메시지를 읽는** 절입니다. 예고와 실제가 다르면

- 「두 개가 동시에 나온다」고 읽고 **메시지를 두 개 찾다가** 하나만 있어 자기가 잘못 실행했다고 의심합니다.
- 「메시지가 길어집니다」라고 했는데 한 줄이라 **다른 에러가 난 줄로 오해**합니다.

실제 메시지(`Target size ... must be the same as input size`)는 오히려 **짧고 명확해서 교육적**입니다. 힌트가 그 장점을 가리고 있습니다.

## Ch 13 #108 과 같은 유형입니다

Ch 13 의 삽질 코너도 「두 에러 동시」를 예고했고 실제와 달라 **#108 로 등록·수정**됐습니다. Ch 17 은 그 문장을 **그대로 물려받았습니다** — 본문도 「Ch 13 의 삽질과 같은 함정」이라고 스스로 밝힙니다.

```bash
# Ch 13 은 고쳐졌고 Ch 17 에만 남아 있습니다
grep -rn "두 가지 에러가 동시에" 13_bert_multilabel/ 17_ko_multilabel/
```

## 기대 결과

힌트를 실제 에러에 맞춥니다. 구조와 어조는 그대로 두고 **예고 부분만** 바꿥니다.

```diff
  힌트: `BCEWithLogitsLoss` 는 *logits 와 같은 shape 의 float 텐서* 를 라벨로 받는데,
  위 코드는 *(B,) int* 를 넘깁니다.
- shape mismatch + dtype mismatch 두 가지 에러가 동시에 날 수 있어 메시지가 길어집니다
- (Ch 13 의 삽질과 같은 함정 — 라벨 *형식* 이 problem_type 과 일치해야 함).
+ `Target size` 와 `input size` 가 다르다는 `ValueError` 가 loss 계산 직전에 납니다 —
+ shape 검사에서 먼저 멈추므로 dtype 문제는 드러나지도 않습니다
+ (Ch 13 의 삽질과 같은 함정 — 라벨 *형식* 이 problem_type 과 일치해야 함).
```

> 실측 메시지를 산문에 그대로 박지 않고 **어떤 종류의 에러가 어디서 나는지**만 적었습니다. 라이브러리 버전에 따라 문구가 달라질 수 있습니다.

## 재현 절차

`repro/ch17_troubleshoot.py` — 챕터 셋업을 최소 재현(klue/bert-base · `multi_label_classification` · 7라벨)하고 `tokenize_wrong` 을 그대로 사용했습니다.

```python
def tokenize_wrong(batch):
    out = tok(batch["text"], truncation=True, max_length=128)
    out["labels"] = [next((i for i, v in enumerate(m) if v > 0), 0) for m in batch["multi_hot"]]
    return out
```

```bash
colab run --gpu T4 --timeout 900 ch17_troubleshoot.py
```

스택 마지막 프레임까지 기록했습니다 — `training_step → compute_loss → model.forward → loss_fct → binary_cross_entropy_with_logits`.

## 전파 범위

| 축 | 상태 |
|---|---|
| 📓 Notebook | c42 **1곳** |
| 📄 WikiDocs | `wrapup.md` **1곳** |
| 📕 EPUB | **해당 없음** — compact 판에 삽질 코너 미수록 |

**재실행 불필요 · 재변환 불필요(마크다운) · tex 무관.**

## 심각도

**Minor** 로 제안합니다. Ch 13 #108 과 같은 등급입니다. 코드·수치는 정상이고 힌트 문구만 실제와 다릅니다. 다만 **독자가 직접 실행하는 절**이라 바로 눈에 띕니다.

## 담당

- 수정 담당자: 미정
- 재검수자: 등록자 (@jwrim9305)
- 연결: 챕터 검수 이슈 #41 / 계열 #108 (Ch 13)

**Comment by jwrim9305** [comment]
## 수정 완료 (@jwrim9305), 2026-09-13

브랜치 `fix/ch17-review-2026-09-b1` @ `f523305`.

```diff
  힌트: `BCEWithLogitsLoss` 는 *logits 와 같은 shape 의 float 텐서* 를 라벨로 받는데,
  위 코드는 *(B,) int* 를 넘깁니다.
- shape mismatch + dtype mismatch 두 가지 에러가 동시에 날 수 있어 메시지가 길어집니다
+ **크기가 먼저 어긋나므로 한 줄짜리 `ValueError` 로 끊깁니다** — shape 검사에서 멈춰
+ dtype 문제는 드러나지도 않습니다
  (Ch 13 의 삽질과 같은 함정 — 라벨 *형식* 이 problem_type 과 일치해야 함).
```

**Ch 13 의 현재 문구와 표현을 맞췄습니다.** 그쪽도 #108 수정으로 「크기가 먼저 어긋나므로 한 줄짜리 `ValueError` 로 끊깁니다」입니다. 같은 함정을 가리키는 두 챕터가 이제 같은 설명을 합니다.

| 축 | 위치 | 재실행 |
|---|---|---|
| 📓 Notebook | `17_ko_multilabel.ipynb` c42 | **불필요** (마크다운) |
| 📄 WikiDocs | `pages/17-ko_multilabel-wrapup.md` | **불필요** |
| 📕 EPUB | **해당 없음** — compact 판에 삽질 코너 미수록 | 불필요 |

「두 가지 에러가 동시에」 잔존 grep **0건** (전 챕터).

> 실측 메시지 문자열을 산문에 박지 않고 **에러 종류와 멈추는 지점**만 적었습니다. 라이브러리 버전에 따라 문구가 달라질 수 있어서입니다. Ch 13 은 메시지 전문을 코드 블록으로 싣고 있는데, 그 방식을 원하시면 맞추겠습니다.

**Comment by jwrim9305** [comment]
## 병합 확인 — 종결합니다 (@jwrim9305), 2026-09-13

**PR #278** 이 `2655899` 로 master 에 병합됐습니다.

| 축 | 확인 | 재실행 |
|---|---|---|
| 📓 Notebook | c42 🚀 삽질 코너 | 불필요 |
| 📄 WikiDocs | `wrapup.md` | 불필요 |
| 📕 EPUB | `ch021.xhtml` — 「한 줄짜리 `ValueError`」 수록 확인 | 불필요 |

「두 가지 에러가 동시에」 잔존 **0건** (전 챕터). Ch 13 #108 수정본과 표현이 일치합니다.

> EPUB 에도 실렸습니다 — compact 판이 삽질 코너를 수록하는 챕터였습니다. 처음 등록 시 「EPUB 해당 없음」으로 적었는데, Ch 16 #158 기준을 그대로 가져온 제 오판이었습니다. 실제로는 📕 에도 반영됐고 확인했습니다.
